Framework/DataHandling/src/SaveNexusGeometry.cpp

```diff
@@ -0,0 +1,57 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidDataHandling/SaveNexusGeometry.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidGeometry/Instrument/InstrumentVisitor.h"
+#include "MantidNexusGeometry/NexusGeometrySave.h"
+
+#include "MantidAPI/FileProperty.h"
+#include "MantidAPI/WorkspaceProperty.h"
+#include "MantidGeometry/Instrument/ComponentInfo.h"
+#include "MantidGeometry/Instrument/DetectorInfo.h"
+
+namespace Mantid {
+namespace DataHandling {
+using Mantid::API::WorkspaceProperty;
+using Mantid::Kernel::Direction;
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(SaveNexusGeometry)
+
+//----------------------------------------------------------------------------------------------
+
+/** Initialize the algorithm's properties.
+ */
+void SaveNexusGeometry::init() {
+  const std::vector<std::string> exts{".nxs", ".hdf5"};
+  declareProperty(std::make_unique<WorkspaceProperty<API::MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input),
+                  "workspace containing the Instrument.");
+
+  declareProperty(std::make_unique<API::FileProperty>(
+                      "Filename", "", API::FileProperty::Save, exts),
+                  "full path save destination");
+
+  declareProperty("H5Path", "", "name of the H5 path save destination group.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void SaveNexusGeometry::exec() {
+  API::MatrixWorkspace_const_sptr workspace = getProperty("InputWorkspace");
+  std::string destinationFile = getPropertyValue("FileName");
+  std::string rootFileName = getPropertyValue("H5Path");
+  const auto instrument = workspace->getInstrument();
```

Comment:
You don't need the instrument visitor. You can get the `ComponentInfo` directly from the input workspace.

`workspace.componentInfo()` similar for `detectorInfo`

Then just call `std::make_pair(std::make_unique(workspace.componentInfo(), workspace.detectorInfo()));` is my suggestion

Reply:
workspace->componentInfo.() and detecttorinfo are private members

Reply:
Really? They look public to me https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/ExperimentInfo.h#L158-L164

Reply:
C2248: 'Mantid::Geometry::ComponentInfo::ComponentInfo': cannot access private member declared in class 'Mantid::Geometry::ComponentInfo'

when calling:

 const std::pair<std::unique_ptr<Geometry::ComponentInfo>,
                  std::unique_ptr<Geometry::DetectorInfo>> &instrument = std::make_pair(std::make_unique<Geometry::ComponentInfo>(compInfo), std::make_unique<Geometry::DetectorInfo>(detInfo));

Reply:
@TakudzwaMakoni 
Ah, that's likely due to copy semantics of componentinfo. You will see from the class defnition that the copy assignment and copy constructor are disabled/private. 

I suggest introducing **NEW** signature (also keep existing for now)

```
MANTID_NEXUSGEOMETRY_DLL void saveInstrument(const Geometry::ComponentInfo&,
                    const Geometry::DetectorInfo& detectorInfo,
    const std::string &fullPath, const std::string &rootPath,
    Kernel::ProgressBase *reporter = nullptr);
```
Since you unpack the pair anyway [here](https://github.com/mantidproject/mantid/blob/master/Framework/NexusGeometry/src/NexusGeometrySave.cpp#L715-L722), make the existing method call the new signature above. And the new signature can be used nicely and bound directly to `workspace.componentInfo()`.